pkg/services/ngalert/state/template.go

```diff
@@ -0,0 +1,73 @@
+package state
+
+import (
+	"context"
+	"math"
+	"strconv"
+	"time"
+
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/template"
+)
+
+// templateCaptureValue represents each value in .Values in the annotations
+// and labels template.
+type templateCaptureValue struct {
+	Labels map[string]string
+	Value  float64
+}
+
+// String implements the Stringer interface to print the value of each RefID
+// in the template via {{ $values.A }} rather than {{ $values.A.Value }}.
+func (v templateCaptureValue) String() string {
+	return strconv.FormatFloat(v.Value, 'f', -1, 64)
+}
+
+func expandTemplate(name, text string, labels map[string]string, alertInstance eval.Result) (result string, resultErr error) {
+	name = "__alert_" + name
+	text = "{{- $labels := .Labels -}}{{- $values := .Values -}}{{- $value := .Value -}}" + text
+	data := struct {
+		Labels map[string]string
+		Values map[string]templateCaptureValue
+		Value  string
+	}{
+		Labels: labels,
+		Values: newTemplateCaptureValues(alertInstance.Values),
+		Value:  alertInstance.EvaluationString,
+	}
+
+	expander := template.NewTemplateExpander(
+		context.TODO(),
+		text,
+		name,
+		data,
+		model.Time(timestamp.FromTime(time.Now())),
+		func(context.Context, string, time.Time) (promql.Vector, error) {
+			return nil, nil
```

Comment:
I'm not sure if we should return `nil, nil` or `nil` and an non-nil error. @gotjosh?

Reply:
I think `nil, nil` is fine - I'd rather proceed with the alert even if the label/annotations is incorrect as opposed to fail to deliver the alert itself.